### PR TITLE
Bug fixes for 2557 and 1567

### LINKED ISSUE #2557: Flaky test: ThreadAffinityTest#testThreadAffinity
### TestNG Version

The latest

### Expected behavior

The test should not fail.

### Actual behavior

https://travis-ci.org/github/cbeust/testng/jobs/772083005#L1270-L1274

```
FAILURE   0.1sec, test.thread.parallelization.ThreadAffinityTest > testThreadAffinity[1]([Ljava.lang.Class;@8402514)
    java.lang.AssertionError: 
    Expected size:<2> but was:<1> in:
    <["5904", "5903"]>
        at test.thread.parallelization.ThreadAffinityTest.testThreadAffinity(ThreadAffinityTest.java:39)
```

@krmahadevan , is there a possibility you remember something regarding the test? ;)

I did try increasing the number of invocations for the test to 10000. Then I configured `LogGatheringListener` to use `Collections.synchronizedSet(Sets.newHashSet())` instead of the regular `Sets.newHashSet()`.

It caused quite a surprising result (one or two among 10000 invocations, Java 15)):

```
Expected size:<2> but was:<3> in:
<["42896", "9448", "42897"]>
java.lang.AssertionError: 
Expected size:<2> but was:<3> in:
<["42896", "9448", "42897"]>
	at test.thread.parallelization.ThreadAffinityTest.testThreadAffinity(ThreadAffinityTest.java:39)
	at jdk.internal.reflect.GeneratedMethodAccessor13.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.testng.internal.invokers.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:136)
	at org.testng.internal.invokers.TestInvoker.invokeMethod(TestInvoker.java:593)
	at org.testng.internal.invokers.TestInvoker.invokeTestMethod(TestInvoker.java:175)
	at org.testng.internal.invokers.MethodRunner.runInSequence(MethodRunner.java:49)
	at org.testng.internal.invokers.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:826)
	at org.testng.internal.invokers.TestInvoker.invokeTestMethods(TestInvoker.java:148)
	at org.testng.internal.invokers.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:147)
	at org.testng.internal.invokers.TestMethodWorker.run(TestMethodWorker.java:129)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.testng.TestRunner.privateRun(TestRunner.java:794)
	at org.testng.TestRunner.run(TestRunner.java:596)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:421)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:415)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:376)
	at org.testng.SuiteRunner.run(SuiteRunner.java:320)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:53)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:96)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1220)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1142)
	at org.testng.TestNG.runSuites(TestNG.java:1071)
	at org.testng.TestNG.run(TestNG.java:1039)
	at org.gradle.api.internal.tasks.testing.testng.TestNGTestClassProcessor.runTests(TestNGTestClassProcessor.java:141)
	at org.gradle.api.internal.tasks.testing.testng.TestNGTestClassProcessor.stop(TestNGTestClassProcessor.java:90)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:61)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:64)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy2.stop(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.stop(TestWorker.java:133)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:64)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:414)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
	at java.base/java.lang.Thread.run(Thread.java:832)
```

### Test case sample

`test.thread.parallelization.ThreadAffinityTest#testThreadAffinity`

### LINKED ISSUE #2567: MethodHelper#CANONICAL_NAME_CACHE is never reset, so it could result in a memory leak
The cache should probably be reworked to a run-scoped or something like that.

### PR DESCRIPTION
**Pull request by krmahadevan**
Closes #2557
Closes #2567

Fixes #2557 #2567  .

### Did you remember to?

- [X] Add test case(s)
- [X] Update `CHANGES.txt`

We encourage pull requests that:

* Add new features to TestNG (or)
* Fix bugs in TestNG

If your pull request involves fixing SonarQube issues then we would suggest that you please discuss this with the 
[TestNG-dev](https://groups.google.com/forum/#!forum/testng-dev) before you spend time working on it.

### CODE REVIEW

File: core/src/main/java/org/testng/internal/MethodHelper.java [hunk 1]
```diff
@@ -456,6 +459,13 @@ protected static String calculateMethodCanonicalName(Class<?> methodClass, Strin
         .orElse(null);
   }
 
+  public static void clear(List<Method> methods) {
+    Optional.ofNullable(methods)
```
Comment (juherr) [inline core/src/main/java/org/testng/internal/MethodHelper.java:463]: List to Stream conversion

I think you can directly use Stream and avoid the intermediate List

File: core/src/main/java/org/testng/TestRunner.java [hunk 2]
```diff
@@ -923,6 +926,17 @@ private void fireEvent(boolean isStart) {
...
+    if (!isStart) {
+      MethodHelper.clear(methods(this.getPassedConfigurations().getAllMethods().stream()));
+      MethodHelper.clear(methods(this.getFailedConfigurations().getAllMethods().stream()));
+      MethodHelper.clear(methods(this.getSkippedConfigurations().getAllMethods().stream()));
+      MethodHelper.clear(methods(Arrays.stream(this.getAllTestMethods())));
+    }
+  }
+
+  private static List<Method> methods(Stream<ITestNGMethod> methods) {
+    return methods.map(each -> each.getConstructorOrMethod().getMethod())
```
Comment (juherr) [inline core/src/main/java/org/testng/TestRunner.java:938]: Stream to List conversion

File: core/src/main/java/org/testng/internal/MethodHelper.java (hunk 1 above)
Comment (krmahadevan) [reply core/src/main/java/org/testng/internal/MethodHelper.java]: fixed

File: core/src/main/java/org/testng/TestRunner.java (hunk 2 above)
Comment (krmahadevan) [reply core/src/main/java/org/testng/TestRunner.java]: fixed

**Comment by krmahadevan** [comment]
@juherr - i have pushed in the amended changes. please take a look

**Review by juherr — APPROVED** [review]

**Pull request merged**